### PR TITLE
Render errors as JSON in http response

### PR DESCRIPTION
**Pull request by newhoggy**
# Description

# Checklist

- [ ] Commit sequence broadly makes sense and commits have useful messages
- [ ] New tests are added if needed and existing tests are updated.  These may include:
  - golden tests
  - property tests
  - roundtrip tests
  - integration tests
  See [Runnings tests](https://github.com/input-output-hk/cardano-node-wiki/wiki/Running-tests) for more details
- [ ] Any changes are noted in the `CHANGELOG.md` for affected package
- [ ] The version bounds in `.cabal` files are updated
- [ ] CI passes. See note on CI.  The following CI checks are required:
  - [ ] Code is linted with `hlint`.  See `.github/workflows/check-hlint.yml` to get the `hlint` version
  - [ ] Code is formatted with `stylish-haskell`.  See `.github/workflows/stylish-haskell.yml` to get the `stylish-haskell` version
  - [ ] Code builds on Linux, MacOS and Windows for `ghc-8.10.7` and `ghc-9.2.7`
- [ ] Self-reviewed the diff

# Note on CI
If your PR is from a fork, the necessary CI jobs won't trigger automatically for security reasons.
You will need to get someone with write privileges.  Please contact IOG node developers to do this
for you.

### CODE REVIEW

File: .github/workflows/check-stylish-haskell.yml [hunk 1]
```diff
@@ -32,16 +32,7 @@ jobs:
...
-        cardano-testnet
-        cardano-git-rev
-        cardano-node
-        cardano-node-chairman
-        cardano-tracer
-        trace-resources
-        cardano-node-capi
-        trace-dispatcher
-        trace-forward
-        cardano-submit-api
```
Comment (smelc) [inline .github/workflows/check-stylish-haskell.yml:44]: Is this change intended?

Comment (newhoggy) [reply .github/workflows/check-stylish-haskell.yml]: I've backed out this change.  Thanks!

File: cardano-testnet/test/cardano-testnet-test/files/golden/tx.failed.response.yaml.golden [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+      - contents:
+          contents: AlonzoInBabbageUtxoPredFailure (BadInputsUTxO (fromList [TxIn
+            (TxId {unTxId = SafeHash "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx"})
+            (TxIx 0)]))
+          tag: UtxoFailure
+        tag: UtxowFailure
+      kind: ShelleyTxValidationError
+    tag: TxValidationErrorInCardanoMode
+  tag: TxCmdTxSubmitValidationError
+tag: TxSubmitFail
```
Comment (newhoggy) [inline cardano-testnet/test/cardano-testnet-test/files/golden/tx.failed.response.yaml.golden:21]: This is the response that will be returned.  The golden file is yaml, but the response will be the JSON equivalent.

File: cardano-submit-api/src/Cardano/TxSubmit/Types.hs [hunk 3]
```diff
@@ -12,36 +16,39 @@ module Cardano.TxSubmit.Types
...
 newtype TxSubmitPort = TxSubmitPort Int
 
 -- | The errors that the raw CBOR transaction parsing\/decoding functions can return.
 --
 newtype RawCborDecodeError = RawCborDecodeError [DecoderError]
-  deriving (Eq, Show)
+  deriving (Eq, Generic, Show)
 
 instance Error RawCborDecodeError where
   prettyError (RawCborDecodeError decodeErrors) = "RawCborDecodeError decode error: " <> pshow (fmap pshow decodeErrors)
```
Comment (carbolymer) [inline cardano-submit-api/src/Cardano/TxSubmit/Types.hs:48]: I think this instance can be simplified a bit
```suggestion
  prettyError (RawCborDecodeError decodeErrors) = "RawCborDecodeError decode error:" <+> pshow decodeErrors
```

**Review by carbolymer — APPROVED** [review]

File: cardano-submit-api/src/Cardano/TxSubmit/Types.hs [hunk 4]
```diff
@@ -50,34 +57,40 @@ data TxSubmitWebApiError
   | TxSubmitBadTx !Text
   | TxSubmitFail TxCmdError
 
-newtype EnvSocketError = CliEnvVarLookup Text deriving (Eq, Show)
+deriving instance Generic TxSubmitWebApiError
```
Comment (Jimbo4350) [inline cardano-submit-api/src/Cardano/TxSubmit/Types.hs:60]: I'd rather we take the time to be explicit about `ToJSON TxSubmitWebApiError`. It gives us better control over deprecation handling or in general customizing the output to our liking.

File: cardano-submit-api/src/Cardano/TxSubmit/Types.hs [hunk 5]
```diff
@@ -50,34 +57,40 @@ data TxSubmitWebApiError
...
+  deriving (Eq, Generic, Show)
+
+instance ToJSON EnvSocketError where
+  toJSON (CliEnvVarLookup msg) = Aeson.object
+    [ "message" .= String msg
+    ]
 
 data TxCmdError
   = TxCmdSocketEnvError EnvSocketError
+  | TxCmdEraConsensusModeMismatch !AnyCardanoEra
```
Comment (Jimbo4350) [inline cardano-submit-api/src/Cardano/TxSubmit/Types.hs:74]: This constructor is not used.

File: cardano-testnet/test/cardano-testnet-test/files/golden/tx.failed.response.yaml.golden (hunk 2 above)
Comment (Jimbo4350) [reply cardano-testnet/test/cardano-testnet-test/files/golden/tx.failed.response.yaml.golden:21]: What's the benefit of yaml vs pretty JSON?

**Review by Jimbo4350 — CHANGES_REQUESTED** [review]
Coupe of minor comments

**Comment by newhoggy** [comment]
> What's the benefit of yaml vs pretty JSON? My thinking is it's better to be close to the actual output.

I switch to use pretty JSON.

YAML would have some benefit if the JSON string contained newlines, but that might not happen in this case.

File: cardano-submit-api/src/Cardano/TxSubmit/Types.hs (hunk 3 above)
Comment (newhoggy) [reply cardano-submit-api/src/Cardano/TxSubmit/Types.hs:48]: This instance wasn't added in the PR.

File: cardano-submit-api/src/Cardano/TxSubmit/Types.hs (hunk 4 above)
Comment (newhoggy) [reply cardano-submit-api/src/Cardano/TxSubmit/Types.hs:60]: I can certainly expand this into a manual `ToJSON` instance.  Do you want the `ToJSON` instance to behave like the generic instance or something different?

File: cardano-submit-api/src/Cardano/TxSubmit/Types.hs (hunk 5 above)
Comment (newhoggy) [reply cardano-submit-api/src/Cardano/TxSubmit/Types.hs]: Removed.  Thanks!

File: cardano-submit-api/src/Cardano/TxSubmit/Types.hs (hunk 4 above)
Comment (Jimbo4350) [reply cardano-submit-api/src/Cardano/TxSubmit/Types.hs:59]: I don't have a preference as long as it's explicit

**Review by Jimbo4350 — APPROVED** [review]
